### PR TITLE
Add CACHE_SIZE environment variable to the available options

### LINKED ISSUE #648: Please allow overriding cache-size
This is a: **FILL ME IN**  
When setting up Pihole along with Unbound, it is best to disable DNSSEC and also set cache to 0, but restarting the Container sets the cache back to 10000.

## Related Issues
- [ ] I have searched this repository/Pi-hole forums for existing issues and pull requests that look similar 

## How to reproduce the issue 

1. Environment data
  * Operating System: **ENTER HERE** 
  * Hardware: 
  * Kernel Architecture: 
  * Docker Install Info and version: 
    - Software source: 
    - Supplimentary Software: 
  * Hardware architecture: 

2. docker-compose.yml contents, docker run shell command, or paste a screenshot of any UI based configuration of containers here
3. any additional info to help reproduce

## These common fixes didn't work for my issue

- [ ] I have tried removing/destroying my container, and re-creating a new container
- [ ] I have tried fresh volume data by backing up and moving/removing the old volume data
- [ ] I have tried running the stock `docker run` example(s) in the readme (removing any customizations I added)
- [ ] I have tried a newer or older version of Docker Pi-hole (depending what version the issue started in for me)
- [ ] I have tried running without my volume data mounts to eliminate volumes as the cause

If the above debugging / fixes revealed any new information note it here.
Add any other debugging steps you've taken or theories on root cause that may help.

### PR DESCRIPTION
**Pull request by jparenas**
## Description

This PR fixes #648 by adding a new environmental variable: CACHE_SIZE. It defaults to 10000 (following the default value in PiHole) and can be changed to any positive number, including 0. This will allow for people to change the default cache size and let the change survive restarts (and avoiding messing with the config file directly).

## Motivation and Context

#648 

## How Has This Been Tested?

I manually tested the change and it works, and I added some test code following some tests that were already present.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x ] My code follows the code style of this project.
- [x ] My change requires a change to the documentation.
- [x ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by dschaper** [comment]
Have you tested a cache size of 0 with DNSSEC enabled?

**Comment by jparenas** [comment]
I haven't tested a cache size of 0 with DNSSEC enabled, but going from the multiple threads online it seems that it would not break anything. 

Also, while I have you here, I saw that the tests are broken as they depend on python:buster and it recently was moved from Python 3.8 to Python 3.9, so they fail on line 20 of Dockerfile_build. Should I submit a new pull request to fix that or can I do it here?

**Comment by dschaper** [comment]
>  it seems that it would not break anything.

Make no assumptions. Test it and verify. `dnsmasq` reacts very badly to 0 cache and DNSSEC. I don't remember offhand if `dnsmasq` now automatically caches or if it still just refuses to start.

I don't know the testing, this isn't my main area but @lightswitch05 or @diginc should be able to help out.

**Comment by jparenas** [comment]
You are absolutely right, dnsmasq refuses to start if cache-size is changed. I made a change to avoid it changing if DNSSEC is enabled and added a test for it.

File: Dockerfile_build [hunk 1]
```diff
@@ -17,7 +17,7 @@ COPY Pipfile* /root/
 WORKDIR /root
 
 RUN pipenv install --system \
-    && sed -i 's|/bin/sh|/bin/bash|g' /usr/local/lib/python3.8/site-packages/testinfra/backend/docker.py
+    && sed -i 's|/bin/sh|/bin/bash|g' /usr/local/lib/python3.9/site-packages/testinfra/backend/docker.py
```
Comment (dschaper) [inline Dockerfile_build:20]: This needs to be set back to 3.8, the `DOCKERFILE_build` in the dev branch is now pinned to 3.8

File: README.md [hunk 2]
```diff
@@ -118,6 +118,7 @@ There are other environment variables if you want to customize various things in
 | `INTERFACE: <NIC>`<br/> *Advanced/Optional* | The default works fine with our basic example docker run commands.  If you're trying to use DHCP with `--net host` mode then you may have to customize this or DNSMASQ_LISTENING.
 | `DNSMASQ_LISTENING: <local\|all\|NIC>`<br/> *Advanced/Optional* | `local` listens on all local subnets, `all` permits listening on internet origin subnets in addition to local.
 | `WEB_PORT: <PORT>`<br/> *Advanced/Optional* | **This will break the 'webpage blocked' functionality of Pi-hole** however it may help advanced setups like those running synology or `--net=host` docker argument.  This guide explains how to restore webpage blocked functionality using a linux router DNAT rule: [Alternative Synology installation method](https://discourse.pi-hole.net/t/alternative-synology-installation-method/5454?u=diginc)
+| `CACHE_SIZE: <size>`<br/> *Advanced/Optional: Default: '10000'* | Set the cache size for dnsmasq. Useful for increasing the default cache size or to set it to 0.
```
Comment (dschaper) [inline README.md:121]: Need to have a note in here that you can not set it to 0 when using DNSSEC.

**Review by dschaper — DISMISSED** [review]

File: Dockerfile_build (hunk 1 above)
Comment (jparenas) [reply Dockerfile_build]: Changed back to 3.8

File: README.md (hunk 2 above)
Comment (jparenas) [reply README.md]: Added the note about DNSSEC. Thanks for bringing it up!

**Comment by jparenas** [comment]
Fixed the requested changes, thanks for bringing them up!

File: bash_functions.sh [hunk 3]
```diff
@@ -193,6 +194,32 @@ setup_dnsmasq_hostnames() {
...
+    local dnsmasq_pihole_01_location="/etc/dnsmasq.d/01-pihole.conf"
+    # Quietly exit early for empty or default
+    if [[ -z "${1}" || "${1}" == '10000' ]] ; then return ; fi
+
+    if [[ "${DNSSEC}" == "true" ]] ; then
+        echo "$warning - Cannot change cache size if DNSSEC is enabled"
+        return
+    fi
+
+    if ! echo $1 | grep -q '^[0-9][0-9]*$' ; then
```
Comment (dschaper) [inline bash_functions.sh:208]: Nitpick: Is there a reason for the regex like this or would `^[0-9]+$` work? Not marking as a change request, but if there's no objection to simplifying it then would be nice to have.

Comment (jparenas) [reply bash_functions.sh]: That's a nice find. I copied this line from https://github.com/pi-hole/docker-pi-hole/blob/f4814935b9c0c531c415b3c9ae58af97466f30cd/bash_functions.sh#L258, but looking closely at it it definitely would be better to simplify it. Changed.

**Comment by jparenas** [comment]
All the requested changes have been made so please let me know if there are any other fixes that should be made to the PR. Thank you for your time and help!

**Comment by PromoFaux** [comment]
@jparenas, apologies for the long time to get back to you. I merged in latest `dev` changes, but it seems that the tests you added are now failing, please take a look

**Comment by jparenas** [comment]
Hi @PromoFaux, sorry for taking a while to get back to you. I updated the branch with a working version (apparently the upstream PiHole install script did some changes that shadowed the CACHE_SIZE env variable. I had to change the variable name because of that. That took a while to figure out) and rebased the branch to the latest commit.

Thank you for taking a look at it after so long!

**Review by PromoFaux — APPROVED** [review]

**Review by PromoFaux — APPROVED** [review]
Re-approve after final sanity check before merge

**Pull request merged**